File: dataset/XGboost.R
#loading data
data <- read.csv("bank-full.csv", sep = ";")

#look at the data structure
str(data)

#creating dataset with numerical variables only
#install.packages("dplyr")
library(dplyr)
dataset_numerical <- data %>% select_if(is.numeric)

#summary statistics and correlation matrix
summary(dataset_numerical)
cor(dataset_numerical)

#adding dependent variable to dataset
dataset_numerical <- cbind(dataset_numerical, data$y)
colnames(dataset_numerical)[8] <- "yes"

#splitting dataset into training and test set
#install.packages("caTools")
library(caTools)
set.seed(1502)
split <- sample.split(dataset_numerical$yes, SplitRatio = 0.8)
training_set <- subset(dataset_numerical, split == TRUE)
test_set <- subset(dataset_numerical, split == FALSE)

#isolating y variable
train.y <- as.numeric(as.factor(training_set$yes)) - 1
test.y <- as.numeric(as.factor(test_set$yes)) - 1

#isolate the x variable
train.X <- as.matrix(training_set[, 2:ncol(training_set)-1])
test.X <- as.matrix(test_set[, 2:ncol(test_set)-1])

#state the parameters
#install.packages("xgboost")
library(xgboost)
parameters <- list(eta = 0.3,
                   max_depth = 2,
                   subsample = 1,
                   colsample_bytree = 1,
                   min_child_weight = 1,
                   gamma = 0,
                   set.seed = 1502, 
                   eval_metric = "auc",
                   objective = "binary:logistic",
                   booster = "gbtree")

#detecting how many cores we haves
#install.packages("doParallel")
library(doParallel)
detectCores()

#Running XGBoost
model1 <- xgboost(data = train.X,
                  label = train.y,
                  set.seed(1502),
                  nthread = 6,
                  nround = 200,
                  params = parameters,
                  print_every_n = 50,
                  early_stopping_rounds = 20,
                  verbose = 1)

#Predictions
predictions1 <- predict(model1, newdata = test.X)
predictions1 <- ifelse(predictions1 > 0.5, 1, 0)

#Checking accuracy
#install.packages("caret")
library(caret)
confusionMatrix(table(predictions1, test.y))
mean(test.y)

############################################################

#tranform the factors into dummy variables
#install.packages("fastDummies")
library(fastDummies)
dataset_dummy <- dummy_cols(data, remove_first_dummy = TRUE)

#get only the dummy variables
dataset_dummy <- dataset_dummy[, 18:ncol(dataset_dummy)]

#creating final dataset
dataset <- cbind(dataset_dummy, dataset_numerical)
dataset <- dataset %>% select(-y_yes)

#############################################################

#splitting dataset into training and test set part 2
library(caTools)
set.seed(1502)
split <- sample.split(dataset$yes, SplitRatio = 0.8)
training_set <- subset(dataset, split == TRUE)
test_set <- subset(dataset, split == FALSE)

#isolating y variable part 2
train.y <- as.numeric(as.factor(training_set$yes)) - 1
test.y <- as.numeric(as.factor(test_set$yes)) - 1

#isolate the x variable par 2
train.X <- as.matrix(training_set[, 2:ncol(training_set)-1])
test.X <- as.matrix(test_set[, 2:ncol(test_set)-1])

#Running XGBoost part 2
library(xgboost)
model2 <- xgboost(data = train.X,
                  label = train.y,
                  set.seed(1502),
                  nthread = 6,
                  nround = 200,
                  params = parameters,
                  print_every_n = 50,
                  early_stopping_rounds = 20,
                  verbose = 1)

#Predictions part 2
predictions2 <- predict(model2, newdata = test.X)
predictions2 <- ifelse(predictions2 > 0.5, 1, 0)

#Checking accuracy
#install.packages("caret")
library(caret)
confusionMatrix(table(predictions2, test.y))

##################################################################
#start parallel processing
library(doParallel)
cpu <- makeCluster(6)
registerDoParallel(cpu)

#state inputs
y <- as.factor(as.numeric(as.factor(dataset$yes)) - 1)
X <- as.matrix(dataset[, 2:ncol(dataset)-1])

#cross validation parameters
library(caret)
tune_control <- trainControl(method = "cv",
                             number = 5,
                             allowParallel = TRUE)

#setting the parameter grid
tune_grid <- expand.grid(nrounds = seq(200, to = 1800, by = 200),
                         eta = c(0.05, 0.3),
                         max_depth = c(2,4,6,8),
                         subsample = c(0.9, 1),
                         colsample_bytree = 1,
                         min_child_weight = 1,
                         gamma = 0)

#cross validation and parameter tuning start
start <- Sys.time()
xgb_tune <- train(x = X,
                  y = y,
                  method = "xgbTree",
                  trControl = tune_control,
                  tuneGrid = tune_grid)
end <- Sys.time()

#retrieve the results
xgb_tune$bestTune


#######################################################################
#start parallel processing
library(doParallel)
cpu <- makeCluster(6)
registerDoParallel(cpu)

#setting the parameter grid part 2
tune_grid2 <- expand.grid(nrounds = seq(200, to = 1800, by = 200),
                         eta = xgb_tune$bestTune$eta,
                         max_depth = xgb_tune$bestTune$max_depth,
                         subsample = xgb_tune$bestTune$subsample,
                         colsample_bytree = c(0.5, 1),
                         min_child_weight = seq(1, to = 4, by = 1),
                         gamma = c(0, 0.05))

#cross validation and parameter tuning start part 2
start <- Sys.time()
xgb_tune2 <- train(x = X,
                  y = y,
                  method = "xgbTree",
                  trControl = tune_control,
                  tuneGrid = tune_grid2)
end <- Sys.time()

#retrieve the results part 2
xgb_tune2$bestTune

#######################################################################

#set the tuned parameters
parameters_tune <- list(eta = xgb_tune2$bestTune$eta,
                   max_depth = xgb_tune2$bestTune$max_depth,
                   subsample = xgb_tune2$bestTune$subsample,
                   colsample_bytree = xgb_tune2$bestTune$colsample_bytree,
                   min_child_weight = xgb_tune2$bestTune$min_child_weight,
                   gamma = xgb_tune2$bestTune$gamma,
                   set.seed = 1502, 
                   eval_metric = "auc",
                   objective = "binary:logistic",
                   booster = "gbtree")

#running xgboost for the final time
model3 <- xgboost(data = train.X,
                  label = train.y,
                  set.seed(1502),
                  nthread = 6,
                  nround = xgb_tune2$bestTune$nrounds,
                  params = parameters_tune,
                  print_every_n = 50,
                  early_stopping_rounds = 20,
                  verbose = 1)

#Predictions part 3
predictions3 <- predict(model3, newdata = test.X)
predictions3 <- ifelse(predictions3 > 0.05, 1, 0)

#Checking accuracy
confusionMatrix(table(predictions3, test.y))

#importance drivers
importance <- xgb.importance(feature_names = colnames(test.X),
                             model = model3)
xgb.plot.importance(importance_matrix = importance)

#shap values
xgb.plot.shap(data = test.X,
              model = model3,
              top_n = 3)

File: dataset/XG_boost.py
#import libraries
import pandas as pd
import xgboost as xgb
import numpy as np

#importing dataset
dataset = pd.read_csv("bank-full.csv", sep = ";")
dataset.dtypes

#isolate the x and y variables
y = dataset.iloc[:, -1].values
X = dataset._get_numeric_data()
    
#split dataset into training and test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size =0.2,
                                                    random_state=1502)

#transform y factor variables
y_train = np.where(y_train == "yes", 1, 0)
y_test = np.where(y_test == "yes", 1, 0)
np.mean(y_train)
np.mean(y_test)

#create xgboost matrices
Train = xgb.DMatrix(X_train, label = y_train)
Test = xgb.DMatrix(X_test, label = y_test)

#set the parameters
parameters1 = {'learning_rate': 0.3,
               'max_depth': 2,
               'colsample_bytree': 1,
               'subsample': 1,
               'min_child_weight': 1,
               'gamma': 0, 
               'random_state': 1502,
               'eval_metric': "auc",
               'objective': "binary:logistic"}

#run XGBoost
model1 = xgb.train(params = parameters1,
                   dtrain = Train,
                   num_boost_round = 200,
                   evals = [(Test, "Yes")],
                   verbose_eval = 50)

#PRedictions
predictions1 = model1.predict(Test)
predictions1 = np.where(predictions1 > 0.5, 1, 0)

#Confusion MAtrix
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report
confusion_matrix1 = confusion_matrix(y_test, predictions1)
print(confusion_matrix1)
report1 = classification_report(y_test, predictions1)
print(report1)

############################################################

#isolate the categorical variables
dataset_categorical = dataset.select_dtypes(exclude = "number")

#tranform categorical variables into dummy variables
dataset_categorical = pd.get_dummies(data = dataset_categorical,
                                     drop_first = True)

#joining numerical and categorical datasets
final_dataset = pd.concat([X, dataset_categorical], axis = 1)

#getting names of columns
feature_columns = list(final_dataset.columns.values)
feature_columns = feature_columns[:-1]

###############################################################

#isolate the x and y variables part 2
y = final_dataset.iloc[:, -1].values
X = final_dataset.iloc[:, :-1].values
    
#split dataset into training and test set
X_train, X_test, y_train, y_test = train_test_split(X, y,
                                                    test_size =0.2,
                                                    random_state=1502)

#create xgboost matrices part 2
Train = xgb.DMatrix(X_train, label = y_train, feature_names = feature_columns)
Test = xgb.DMatrix(X_test, label = y_test, feature_names = feature_columns)

#set the parameters part 2
parameters2 = {'learning_rate': 0.3,
               'max_depth': 2,
               'colsample_bytree': 1,
               'subsample': 1,
               'min_child_weight': 1,
               'gamma': 0, 
               'random_state': 1502,
               'eval_metric': "auc",
               'objective': "binary:logistic"}

#run XGBoost
model2 = xgb.train(params = parameters2,
                   dtrain = Train,
                   num_boost_round = 200,
                   evals = [(Test, "Yes")],
                   verbose_eval = 50)

#Predictions part 2
predictions2 = model2.predict(Test)
predictions2 = np.where(predictions2 > 0.5, 1, 0)

#Confusion MAtrix
confusion_matrix2 = confusion_matrix(y_test, predictions2)
print(confusion_matrix2)
report2 = classification_report(y_test, predictions2)
print(report2)

################################################################

#checking how many cores we have
import multiprocessing
multiprocessing.cpu_count()

#setting the cross validation parameters
from sklearn.model_selection import KFold
tune_control = KFold(n_splits = 5,
                     shuffle = True,
                     random_state = 1502).split(X = X_train,
                                                y = y_train)
                                                
#set parameter tuning
#set the parameters part 2
tune_grid = {'learning_rate': [0.05, 0.3],
               'max_depth': range(2, 9, 2),
               'colsample_bytree': [0.5, 1],
               'subsample': [1],
               'min_child_weight': [1],
               'gamma': [0], 
               'random_state': [1502],
               'n_estimators': range(200, 2000, 200),
               'booster': ["gbtree"]}                                                

#State that we are doing a classification problem
from xgboost import XGBClassifier
classifier = XGBClassifier(objective = "binary:logistic")

#Cross Validation Assembly
from sklearn.model_selection import GridSearchCV
grid_search = GridSearchCV(estimator = classifier,
                           param_grid = tune_grid,
                            scoring = "roc_auc",
                            n_jobs = 6,
                            cv = tune_control,
                            verbose = 5)

#Setting evaluation parameters
evaluation_parameters = {"early_stopping_rounds": 100,
                         "eval_metric": "auc",
                         "eval_set": [(X_test, y_test)]}

#Hyperparameter tuning and cross validation
tune_model = grid_search.fit(X = X_train,
                             y = y_train,
                             **evaluation_parameters)
grid_search.best_params_, grid_search.best_score_

#############################################################

#setting the cross validation parameters
from sklearn.model_selection import KFold
tune_control = KFold(n_splits = 5,
                     shuffle = True,
                     random_state = 1502).split(X = X_train,
                                                y = y_train)
                                                
#set parameter tuning part 2
tune_grid2 = {'learning_rate': [0.05],
               'max_depth': [6],
               'colsample_bytree': [0.5],
               'subsample': [0.9, 1],
               'min_child_weight': range(1,5,1),
               'gamma': [0, 0.1], 
               'random_state': [1502],
               'n_estimators': range(200, 2000, 200),
               'booster': ["gbtree"]}                                                

#Cross Validation Assembly
from sklearn.model_selection import GridSearchCV
grid_search2 = GridSearchCV(estimator = classifier,
                           param_grid = tune_grid2,
                            scoring = "roc_auc",
                            n_jobs = 6,
                            cv = tune_control,
                            verbose = 5)

#Hyperparameter tuning and cross validation
tune_model2 = grid_search2.fit(X = X_train,
                             y = y_train,
                             **evaluation_parameters)
grid_search2.best_params_, grid_search2.best_score_

##########################################################

#set the parameters part 3
parameters3 = {'learning_rate': 0.05,
               'max_depth': 6,
               'colsample_bytree': 0.5,
               'subsample': 1,
               'min_child_weight': 2,
               'gamma': 0, 
               'random_state': 1502,
               'eval_metric': "auc",
               'objective': "binary:logistic"}

#run XGBoost part 3
model3 = xgb.train(params = parameters3,
                   dtrain = Train,
                   num_boost_round = 800,
                   evals = [(Test, "Yes")],
                   verbose_eval = 50)

#Predictions part 3
predictions3 = model3.predict(Test)
predictions3 = np.where(predictions3 > 0.05, 1, 0)

#Confusion Matrix
confusion_matrix3 = confusion_matrix(y_test, predictions3)
print(confusion_matrix3)
report3 = classification_report(y_test, predictions3)
print(report3)

Predictions  = 9041
Predictions_no = 7651 + 599
predictions_yes= 502 + 291

#plot importances
xgb.plot_importance(model3, max_num_features = 10)

#Preparing SHAP
#pip install shap
#conda install -c conda-forge shap
import shap
explainer = shap.TreeExplainer(model3)
shap_values = explainer.shap_values(X_test)

#shap plotting
shap.summary_plot(shap_values,
                  X_test,
                  feature_names = feature_columns,
                  max_display = 10)

























